mir: init at 2.11.0

onny opened the pull request:
###### Description of changes

Packaging Canonicals display server Mir https://github.com/MirServer/mir

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

onny committed:
mir: init at 2.11.0

onny committed:
eglexternalplatform: init at 1.1

AndersonTorres commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,111 @@
+    python39Packages.pillow
+    udev
+    wayland
+    xorg.libX11
```
xorg libraries should be declared directly, without `xorg.` redirection.

SuperSandro2000 replied:
It is actually better doing it like this because inheriting from top-level breaks splicing.

AndersonTorres replied:
Well, why not using `pkgs` then? As far as I remember, splicing is problematic only in very special cases, and it is fixed by explicit decoration like `splicedPackages`, not by a magic choice of "should I use xorg.libxcb or libxcb?".

AndersonTorres commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,111 @@
+    license = licenses.gpl2Plus;
+    maintainers = with maintainers; [ onny ];
+  };
+
```
```suggestion
```

AndersonTorres commented:
Somewhat related:

https://github.com/NixOS/nixpkgs/issues/99090

AndersonTorres commented on pkgs/development/libraries/eglexternalplatform/default.nix:
```diff
@@ -0,0 +1,35 @@
+, fetchFromGitHub
+}:
+
+stdenv.mkDerivation rec {
```
Does this package need a C compiler? If not, then use stdenvNoCC

SuperSandro2000 commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,111 @@
+    xorg.xorgproto
+  ];
+
+  # Tests depend on further packages which are not yet packaged
```
which ones?

SuperSandro2000 commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,111 @@
+    nettle
+    pcre
+    pcre2
+    python39Packages.pillow
```
```suggestion
    python3Packages.pillow
```

SuperSandro2000 commented on pkgs/development/libraries/eglexternalplatform/default.nix:
```diff
@@ -0,0 +1,35 @@
+  src = fetchFromGitHub {
+    owner = "Nvidia";
+    repo = "eglexternalplatform";
+    rev = "7c8f8e2218e46b1a4aa9538520919747f1184d86";
```
```suggestion
    rev = version;
```

SuperSandro2000 commented:
This PR rebuilds a lot of packages which means we must target staging. Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging) to not potentially ping a lot of people.

OPNA2608 reviewed (changes requested):
Thanks for looking into continuing the Lomiri stuff :smile:! Some feedback on the Mir packaging here:

It's not of a high priority but there may be some docs that could be built:
```
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) (Required is at least version "1.8.0")
```

Some more patching seems to be needed:

```
[ 27%] Generating drm-formats
grep: /var/empty/include/libdrm/drm_fourcc.h: No such file or directory
[ 28%] Generating drm-formats-big-endian
grep: /var/empty/include/libdrm/drm_fourcc.h: No such file or directory
```
I think these paths here need to be patched to go somewhere into `libdrm`: https://github.com/MirServer/mir/blob/2a130a264fb68188ea6ea699744e32ab7b25432c/src/platform/graphics/CMakeLists.txt#L51-L95. Trying to run the demos errors on some DRM stuff for me, might be related to this.

```
-- Installing: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/graphics-gbm-kms.so.20
-- Set runtime path of "/nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/graphics-gbm-kms.so.20" to ""
-- Installing: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/server-x11.so.20
-- Installing: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/graphics-eglstream-kms.so.20
-- Installing: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/graphics-wayland.so.20
-- Installing: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/input-evdev.so.8
```
```
[2022-12-25 23:06:34.802902] <information> mircommon: Loading modules from: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0//nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform
[2022-12-25 23:06:34.803228] <information> mircommon: Loading module: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0//nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/graphics-gbm-kms.so.20
[2022-12-25 23:06:34.803422] <information> mircommon: Loading module: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0//nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/server-x11.so.20
[2022-12-25 23:06:34.803509] <information> mircommon: Loading module: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0//nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/graphics-eglstream-kms.so.20
[2022-12-25 23:06:34.803581] <information> mircommon: Loading module: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0//nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/graphics-wayland.so.20
[2022-12-25 23:06:34.803644] <information> mircommon: Loading module: /nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0//nix/store/rq0nihx42jqc7sc7dw4gm99y0h15ak40-mir-2.10.0/lib/mir/server-platform/input-evdev.so.8
```
Not sure where this needs to be patched but `/nix/store/.../nix/store/...`-y paths look like something's broken.

onny commented:
Thank you so far for reviewing and helping with this PR :)

@OPNA2608 I fixed the issues you mentioned, thank you for noticing!

OPNA2608 commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,121 @@
+
+  nativeBuildInputs = [
+    cmake
+    doxygen
```
```
-- doxygen 1.9.5 (>= 1.8.0) available - enabling make target doc
```

Doesn't seem enough to actually build & install any documentation, you may have to play with either the `makeFlags` to build & install all normal targets + docs or add some `postBuild` & `postInstall` code to do it.

OPNA2608 replied:
`libxslt` in `nativeBuildInputs` and
```nix
buildFlags = [ "all" "doc" ];
```

will build & install the documentation.

![Bildschirmfoto von 2022-12-27 16-54-56](https://user-images.githubusercontent.com/23431373/209690520-37825291-69ab-435e-94a1-c1b882f140fa.png)

OPNA2608 commented:
I don't have any knowledge about Mir so I can't judge if everything works & is configured correctly, but it seems like a demo atop your current display server with your preferred terminal emulator can be started with the following (replace `kitty` with whatever terminal you wish to use ofc, YMMV):
```sh
miral-app -kiosk -terminal kitty
```

FWIW it seems to launch:

https://user-images.githubusercontent.com/23431373/209588012-2e998a4c-7f65-4e21-b734-9f1f5e65350d.mp4

OPNA2608 commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,121 @@
+    homepage = "https://github.com/MirServer/mir";
+    license = licenses.gpl2Plus;
+    maintainers = with maintainers; [ onny ];
+  };
```
https://discourse.ubuntu.com/t/mir-graphics-support/13185#heading--summary-mir-graphics-platforms
![Bildschirmfoto von 2022-12-27 14-42-40](https://user-images.githubusercontent.com/23431373/209675323-c1675669-99e0-4520-bf9b-8b1ce91009e8.png)

I think all these platforms boil down to Linux-only, more or less?

```suggestion
    platforms = platforms.linux;
  };
```

onny commented:
Thank you so much for the feedback :) Added the requested changes

OPNA2608 reviewed (changes requested):
The order of commits should be fixed: First `eglexternalplatform`, then `mir`.

Starting to run out of suggestions for things that could be improved, I promise! :smile:

Some notes on all the Mir stuff, after toying around with it abit more:

---

The `miral-app` script wants `inotify-tools` so it can nicely wait for things to appear in the background. I don't know if this script matters to how Lomiri will be launched, might make sense to patch it in:

```
↪ /nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app -kiosk -terminal kitty
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
/nix/store/wa9fkqzb10zfxlymg11z115mfb058fyk-mir-2.10.0/bin/miral-app: line 87: inotifywait: command not found
[2022-12-28 15:44:24.815614] <information> mirserver: Starting
```

---

The `miral-terminal` script is supposed to check for a range of different terminal emulators and launch one as the terminal inside Mir, on my system it can find none.

> `for terminal in gnome-terminal.real gnome-terminal weston-terminal qterminal lxterminal x-terminal-emulator xdg-terminal`

I use none of the named terminals there. `x-terminal-emulator` is Debianism for their alternatives system where you would register your preferred terminal emulator. `xdg-terminal`… I don't know what that is.
`xterm` doesn't launch reliably. In kiosk mode it can't find a `DISPLAY`, likely because it doesn't use Xwayland there. Not sure how this should be handled. For now, `miral-app -terminal <terminal>` lets you override `miral-terminal` with whatever you wish to use instead.

---

(Mir? `miral-shell`?) seems to need a full path to `Xwayland`, else it won't launch. Looks like the default for this is set [here](https://github.com/MirServer/mir/blob/7096f877888496f2eb6144743a511c866b38e460/src/miral/x11_support.cpp#L61-L63). Don't know if we should ultimately just hardcode an `Xwayland` into there or work around this via some more configurable way (if there is one):
```
Mir fatal error: Cannot execute Xwayland: --xwayland-path /usr/bin/Xwayland
Aborted (core dumped)
```

---

It also expects Ubuntu's cursor themes to be present:
```
[2022-12-28 16:49:41.945210] < -warning- > miral: Failed to load cursor theme: default
[2022-12-28 16:49:41.945302] < -warning- > miral: Failed to load cursor theme: DMZ-White
ERROR: /build/source/src/miral/cursor_theme.cpp(74): Throw in function miral::CursorTheme::operator()(mir::Server&) const::<lambda()>
Dynamic exception type: boost::wrapexcept<std::runtime_error>
std::exception::what: Failed to load cursor theme: default:DMZ-White
```

AFAICT we actually have them packaged under `vanilla-dmz`? Or at least we're using the correct source there, but instead of installing them as `DMZ-White` and `DMZ-Black` they're called `Vanilla-DMZ` and `Vanilla-DMZ-AA` respectively… Not sure if there's a reason for that.

As a test I've tried to symlink them to the names they should have, adding them to `XCURSOR_PATH` in a `nix-shell` and using `--xwayland-path`. This seems to make Mir much happier and `miral-shell` usable (via `miral-app` wrapper due to the `miral-terminal` stuff):

https://user-images.githubusercontent.com/23431373/209839480-5064d753-e43b-4dab-880a-ed33752d0ae4.mp4

On a tty as well:

![image](https://user-images.githubusercontent.com/23431373/209848669-a753b2d0-7dd2-45fb-ac1f-371afbb1fcdd.png)

And on aarch64-linux hardware:

![image](https://user-images.githubusercontent.com/23431373/209861649-4a780fe3-c522-415c-8393-7b62d3847af0.png)

onny commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,126 @@
+
+stdenv.mkDerivation rec {
+  pname = "mir";
+  version = "2.10.0";
```
Oh you're right, good point https://github.com/MirServer/mir/issues/1744

I guess we'll have to add Mir 1.X with qtmir in a separate PR. Since Mir can be used outside of qtmir, we would start here with package the most recent version

OPNA2608 commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,126 @@
+    nettle
+    pcre
+    pcre2
+    python3Packages.pillow
```
I looked further into this. I can build it fine with the following dependencies. (Note some added dependencies in `nativeBuildInputs` for cross & `strictDeps = true` to work):
```nix
{
  nativeBuildInputs = [ 
    cmake
    pkg-config
    pythonEnv
    glib # gdbus-codegen
    lttng-ust # lttng-gen-tp
    doxygen
    libxslt
  ];  

  buildInputs = [ 
    boost
    egl-wayland
    freetype
    glib
    glm 
    glog
    libdrm
    libepoxy
    libevdev
    libglvnd
    libinput
    libuuid
    libxcb
    libxkbcommon
    libxmlxx
    libyamlcpp
    lttng-ust
    mesa
    nettle
    udev
    wayland
    xorg.libX11
    xorg.libXcursor
    xorg.xorgproto
  ];
}
```

The following `buildInputs` only silence some warnings from `pkg-config` about dependencies of dependencies being missing. They aren't needed for any re-linking by Mir, they find their way into the final binaries just fine according to `ldd` and the dependencies transitively pull them into the closure anyway, just not into the build environment. I'm slightly in favour of throwing them out so the `buildInputs` *look* abit thinner at the cost of some warnings in the logs but overall I'm impartial to it.

```
    eglexternalplatform
    pcre
    pcre2
    libselinux
    libsepol
    xorg.libXau
    xorg.libXdmcp
    xorg.libXrender
```

OPNA2608 commented:
https://github.com/NixOS/nixpkgs/pull/207534#discussion_r1063721210 about the `nativeBuildInputs` and `buildInputs` stuff aside, the commit order is still wrong. The `mir` commit depends on `eglexternalplatform` before the `eglexternalplatform` commit moved it out of `egl-wayland`.

Those are the only nitpick I can think of, LGTM otherwise. I got `wlcs` packaged on master and Mir's tests working in a branch (which address the Xwayland point from earlier out of necessity), can PR that once this lands.

OPNA2608 commented on pkgs/servers/mir/default.nix:
```diff
@@ -0,0 +1,129 @@
+  };
+
+  postPatch = ''
+    # Fix error broken path found in pc file ("//")
+    for f in $(find . -name '*.pc.in') ; do
+      substituteInPlace $f \
+        --replace "/@CMAKE_INSTALL_LIBDIR@" "@CMAKE_INSTALL_LIBDIR@" \
+        --replace "/@CMAKE_INSTALL_INCLUDEDIR@" "@CMAKE_INSTALL_INCLUDEDIR@"
+    done
```
While trying to build something that depends on Mir we noticed that the paths are still broken, just not broken enough for the pkg-config check to notice:

```
prefix=/nix/store/yr3jc3213w1y3f1kc7ppa1a5yx0xng4y-mir-2.10.0
libdir=${prefix}/nix/store/yr3jc3213w1y3f1kc7ppa1a5yx0xng4y-mir-2.10.0/lib
includedir=/nix/store/vzy7w5yqi6y5z5qjpiyq1yfpjpxkzk6i-mir-2.10.0-dev/nix/store/vzy7w5yqi6y5z5qjpiyq1yfpjpxkzk6i-mir-2.10.0-dev/include/miral
```

I think for our purposes, replacing all `${prefix}/@CMAKE_INSTALL_<dir>@` with `@CMAKE_INSTALL_FULL_<dir>@` etc might be fine:

```
prefix=/nix/store/q4921w10hjmpk28dkk1c4y1c5fy33fcy-mir-2.10.0
libdir=/nix/store/q4921w10hjmpk28dkk1c4y1c5fy33fcy-mir-2.10.0/lib
includedir=/nix/store/ird024yncpxv3rlz42grx8i4g3s891s4-mir-2.10.0-dev/include/miral
```

The more proper fix being something like [jtojnar/cmake-snips#concatenating-paths-when-building-pkg-config-files](https://github.com/jtojnar/cmake-snips/tree/a9d72829ab0cb42e541c9975e698d6b8282610ed#concatenating-paths-when-building-pkg-config-files) being done upstream.

AndersonTorres replied:
> being done upstream.

Should we `@-`mark them?

OPNA2608 replied:
I PR'd the same fix to their wlcs project recently (first with `FULL`, now with jtojnar's snippet), guess I could do the same for Mir & link it here for `fetchpatch`ing.

OPNA2608 replied:
https://github.com/MirServer/mir/pull/2786, should be cherry-pickable onto 2.11.0.

OPNA2608 commented:
New release about a week ago: [v2.11.0](https://github.com/MirServer/mir/releases/tag/v2.11.0)

onny commented:
> New release about a week ago: [v2.11.0](https://github.com/MirServer/mir/releases/tag/v2.11.0)

Thanks for the note, updated it to the latest version and changed the commit order. For the other issues, I'll have to look more into it.
Feel free to push into this PR or continue it separately :)

phossil commented:
Mir provides a wayland session file that can be added with `passthru.providedSessions = [ "mir-shell" ];`.

![20230115_212127](https://user-images.githubusercontent.com/38191182/212588019-c74f7b8c-5200-49f1-aa10-fcd096265d82.jpg)
